### PR TITLE
"episode 4" -- first example recipe

### LINKED ISSUE #29: Episode 4 Running a recipe (First example):
Internal name: first_recipe

The outline document says:
 
- Questions:
  - What is a recipe?
  - How can I do the same preprocessing on many different datasets?
  - What are the files and directories after running a recipe?
  - What happens when I run a recipe?

- Objectives: 	
  - Run an ESMValTool recipe
  - Understand the purpose of different settings in the recipe
  - Inspect the output directories
  - Examine the log information

- Contents:
  - How to run ESMValTool 
  - esmvaltool -c user-config.yml recipe.yml
  - A basic recipe that takes a simple dataset and produces a simple plot.
  - We’ve found that the best example here is a scalar (time series) field, like the thetaoga (global average ocean temperature). This data is small and light, (hundreds of kb per file). 

- Discussion:
  - Inspect the example recipe
  - What do you think it will produce?
  - Exercise: Run the example recipe to produce a simple time series of ~20 years of data.

- Discussion: Inspect the output:
  - Plots
  - Logs
  - Preprocessed netcdf
  - Provenance and citation information
  - Settings.yml and metadata/yml files

- Exercise:
  - edit the file to include other models or a different variable and re-run.

- Exercise:
  - edit the recipe to run:
  - Atmosphere: Temperature at 2m
  - Land surface: Average land surface tempureate in your local region.
  - Ocean: Surface ocean temperature

- Callout: Common issues & tips
  - ESMValTool can’t locate the data (user didn’t correctly edit user-config.yml)
  - Esmvaltool not found (user didn’t active or correctly install conda)
  - Diagnostic path problems (explain ESMValTool’s methods to determine diagnostic path, and how it should appear in the recipe)
  - FX files not found.
  - The preprocessor works but the diagnostic fails. How to tell them appart and how to re-run a failed diagnostic (but not the preprocessor)
  - Your recipe’s name/project/reference isn’t recognised by ESMValTool.

- Key points: 
  - You can’t break A a recipe does not break by fiddling with it
  - Log information is useful to How to interpret the first warnings/errors
  - The dataset section in the recipe relates to Understanding the directory structure of CMIP data on your server/disk and knowing how to use data from different experiments such as CMIP/ScenarioMIP.

### PR DESCRIPTION
**Pull request by BenMGeo**
# Pull Request checklist

We appreciate your time and effort to improve the tutorial. Please keep in mind that lesson maintainers are volunteers and it may be some time before they can respond to your contribution.

* * *

## Before you start

- [x] Read [CONTRIBUTING.md](https://github.com/ESMValGroup/tutorial/blob/master/CONTRIBUTING.md).
- [x] Create an [issue](https://github.com/ESMValGroup/tutorial/issues) to discuss your idea. This allows your contributions to be incorporated into the tutorial.

## Tasks

- [x] Give this pull request a descriptive title.
- [x] If you are contributing to existing lesson materials, please make sure the content conforms to the `Lesson development` section in [CONTRIBUTING.md](https://github.com/ESMValGroup/tutorial/blob/master/CONTRIBUTING.md) and does not contain any spelling or grammatical errors.
- [ ] If you are making a new episode, please make sure the content conforms to the `Lesson organization` and `Lesson formatting` sections in [CONTRIBUTING.md](https://github.com/ESMValGroup/tutorial/blob/master/CONTRIBUTING.md) and does not contain any spelling or grammatical errors.
- [ ] Preferably Codacy checks pass. Status can be seen below your pull request. If there is an error, click the link to find out why.
- [ ] Preview changes on your machine before pushing them to GitHub by running `make serve`, alternatively `make docker-serve`. Please see the `Previewing your changes locally` section in [CONTRIBUTING.md](https://github.com/ESMValGroup/tutorial/blob/master/CONTRIBUTING.md) for installation instructions.
- [x] All code instructions have been tested.

If you need help with any of the tasks above, please do not hesitate to ask by commenting in the issue or pull request.

* * *

Closes #29

### CODE REVIEW

**Comment by ledm** [comment]
Nice work @BenMGeo!

File: _episodes/first_example_recipe.md [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+ - which project is responsible for it,
+ - and which publications, references are linked with the recipe.
+
+ Note that the authors, publications and references are to be named in the config-references.yml
+
+This information ...
+
+## How to run ESMValTool
+
+Once you’ve set up your conda environment and installed ESMValTool (see episode #2 LINK) and set up your config-user.yml file to correctly match you local environment, (see episode #3 LINK), ESMValTool is invoked using a simple command:
```
Comment (ledm) [inline _episodes/first_example_recipe.md:60]: How do we add a link to the other episodes here?

File: _episodes/first_example_recipe.md [hunk 2]
```diff
@@ -0,0 +1,233 @@
+---
+title: "Running a recipe (First example)"
+teaching: 20
+exercises: 25
+questions:
+- "What is a recipe?"
+- "How can I do the same preprocessing on many different datasets?"
+- "What are the files and directories after running a recipe?"
```
Comment (hb326) [inline _episodes/first_example_recipe.md:8]: If it does not make this question too long I would rephrase "What are the files and directories that are created after running a recipe?"

Comment (BenMGeo) [reply _episodes/first_example_recipe.md:8]: changed in last commit

File: _episodes/first_example_recipe.md [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+- "What happens when I run a recipe?"
+objectives:
+- "Run an ESMValTool recipe"
+- "Understand the purpose of different settings in the recipe"
+- "Inspect the output directories"
+- "Examine the log information"
+keypoints:
+- "A recipe does not break by fiddling with it"
+- "Log information is useful when interpreting the first warnings/errors"
+- "The dataset section in the recipe relates to Understanding the directory structure of CMIP data on your server/disk and knowing how to use data from different experiments such as CMIP/ScenarioMIP"
```
[the remark this answers is not in the document]
Comment (BenMGeo) [reply _episodes/first_example_recipe.md]: changed the wording

File: _episodes/first_example_recipe.md (hunk 1 above)
Comment (BenMGeo) [reply _episodes/first_example_recipe.md:61]: https://carpentries.github.io/lesson-example/04-formatting/index.html

File: _episodes/first_example_recipe.md [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+> Use the command:
+> ~~~bash
+> esmvaltool -c ./path_to_file/user-config.yml ./path_to_file/recipe_example.yml
+> ~~~
+>
+> Follow the terminal guiding you through the subprocesses that are running. Can you find where the preprocessor and the diagnostic are starting? Which one took longer to process?
+{: .challenge}
+
+> ## Exemplary output
+> FIXME
```
Comment (BenMGeo) [inline _episodes/first_example_recipe.md:213]: The FIXMEs have to be filled out by someone who can run the recipe on JASMIN (I do not have an account).

It is solutions to the challenges and they should be as close to the expected results as possible.

File: _episodes/first_example_recipe.md [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+> ## ESMValTool can’t locate the data. The error message is `esmvalcore._recipe_checks.RecipeError: Missing data`
+> Which computing machine are you using? Does your user-config.yml file reflect your machine's settings? Is the dataset’s name in the correct order?
+{: .solution}
+
+> ## Diagnostic path problems
+> The directory path to your diagnostics code is set relative to the esmvaltool/diag_scripts subdirectory. Is the code placed in this subdirectory? Is it spelled correctly?
+{: .solution}
+
+> ## FX files not found
+> There is no FX file for your corresponding dataset. Are your datasets’ names spelled correctly? Is there a FX file for this dataset?
```
Comment (BenMGeo) [inline _episodes/first_example_recipe.md:323]: @ledm Is this correct?

**Comment by BenMGeo** [comment]
I will produce the first version of the output for the FIXMEs.

File: _episodes/first_example_recipe.md [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+---
+
+This episode describes how ESMValTool recipes work, how to run a recipe and how to explore the recipe output. By the end of this episode, you should be able to run your first recipe, look at the recipe output, modify a recipe, explore and run some basic recipe debugging.
+
+## Introduction to Recipes
+
+Recipes are the instructions that you give to ESMValTool that tell it what you want to do. This includes four main sections: datasets, preprocessors, diagnostics and description.
+
+  - datasets: what datasets you want to use, including
+    - the time range and time resolution,
```
Comment (hb326) [inline _episodes/first_example_recipe.md:28]: "time period" instead of "time range"?
"temporal resolution" instead of "time resolution"?

File: _episodes/first_example_recipe.md [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+    - the MIP, ensemble member,
```
Comment (hb326) [inline _episodes/first_example_recipe.md:29]: Should "ensemble member" be an extra bullet point?

Comment (hb326) [reply _episodes/first_example_recipe.md:29]: And maybe we have to provide an example for "MIP"?

File: _episodes/first_example_recipe.md [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+    - the experiment (i.e. historical, ssp125, etc.),
+    - and the grid type (CMIP6 only).
```
Comment (hb326) [inline _episodes/first_example_recipe.md:31]: I would rephrase to "(necessary for CMIP6 only)"

File: _episodes/first_example_recipe.md [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+
+  - preprocessors: general operations applied to a dataset before handling it in a diagnostic, listing
```
Comment (hb326) [inline _episodes/first_example_recipe.md:33]: I am not sure about the word "listing" at the end. Maybe use "including" as in the description of the other three parts of the recipe or use "defining"?

File: _episodes/first_example_recipe.md [hunk 10]
```diff
@@ -0,0 +1,332 @@
...
+    - which preprocessor modules to apply,
+    - the order to apply them,
```
Comment (hb326) [inline _episodes/first_example_recipe.md:35]: Rephrase to "the order in which they are applied"?

File: _episodes/first_example_recipe.md [hunk 11]
```diff
@@ -0,0 +1,332 @@
...
+    - and the preprocessor arguments.
+
+    This section can also be optional, if no preprocessing is needed.
+
+  - diagnostics: all the information about the diagnostic, including
+    - list of variables to evaluate (with their respective configurations),
+    - the desired diagnostic script to use,
+    - and additional diagnostic script options or arguments, if needed.
+
+    Also include additional datasets beyond those included in the datasets section mentioned above, for instance variable specific observational data.
```
Comment (hb326) [inline _episodes/first_example_recipe.md:45]: Maybe rephrase to "It is possible to also include additional...."

Comment (hb326) [reply _episodes/first_example_recipe.md:45]: The additional dataset setting has also been discussed in episode 5 (preprocessing).

File: _episodes/first_example_recipe.md [hunk 12]
```diff
@@ -0,0 +1,332 @@
...
+
+  - description: a brief description of the recipe, including
+    - who wrote the recipe and who maintains it,
+    - which project is responsible for it,
```
Comment (hb326) [inline _episodes/first_example_recipe.md:49]: I don't think we can phrase it like this. No project will ever be responsible for a diagnostic. Mabye we could phrase it like this "in which project framework this diagnistic was develped" or "for which project this diagnostic was developed"

File: _episodes/first_example_recipe.md [hunk 13]
```diff
@@ -0,0 +1,332 @@
...
+    - and which publications, references are linked with the recipe.
```
Comment (hb326) [inline _episodes/first_example_recipe.md:50]: I would replace the "," with an "and"

File: _episodes/first_example_recipe.md [hunk 14]
```diff
@@ -0,0 +1,332 @@
...
+
+    Note that the authors, publications and references are to be named in the config-references.yml
```
Comment (hb326) [inline _episodes/first_example_recipe.md:52]: Maybe rephrase "Note that the authors, publications and references need to be included in the config-reference.yml for the recipe to run successfully."

File: _episodes/first_example_recipe.md [hunk 15]
```diff
@@ -0,0 +1,332 @@
...
+
+The information you provide in the recipe is not only affecting the processes you are starting, but also the directory names your output will be structured in.
+For additional reeds, please have a look at the recipe format description in the [ESMValTool manual](https://docs.esmvaltool.org/projects/esmvalcore/en/latest/recipe/overview.html#recipe-section-diagnostics).
```
Comment (hb326) [inline _episodes/first_example_recipe.md:55]: Do you mean "reads" instead of "reeds"?

File: _episodes/first_example_recipe.md [hunk 16]
```diff
@@ -0,0 +1,332 @@
...
+~~~source
+esmvaltool -c configuration recipe
+~~~
+
+To try your hand with a basic recipe, please work through this episode.
+
+## Introduction to the example recipe
+The recipe presented here is a simple, basic recipe that takes a single dataset and produces a time series plot.
+
+Please download the following recipe into your ESMValTool working area with the name: recipe_example.yml LINK
```
Comment (hb326) [inline _episodes/first_example_recipe.md:70]: Maybe rephrase to "... into your ESMValTool working directory..."?

File: _episodes/first_example_recipe.md (hunk 4 above)
Comment (BenMGeo) [reply _episodes/first_example_recipe.md:213]: Based on discussion @workshop, I include it based on my output.

File: _episodes/first_example_recipe.md [hunk 17]
```diff
@@ -0,0 +1,332 @@
...
+> 38            short_name: thetaoga
+> 39            preprocessor: prep_timeseries
+> 40        scripts:
+> 41          timeseries_diag:
+> 42            script: ocean/diagnostic_timeseries.py
+> ~~~
+{: .solution}
+
+> ## Explore the recipe
+> Use the command and investigate the sample recipe.
```
Comment (hb326) [inline _episodes/first_example_recipe.md:120]: Maybe rephrase to "Use the following command...."
But in any case we might want to streamline the suggestions for text editors throughout the episodes so that they are consistent. At the moment I have seen two different editor suggestions, and I have not seen all of the episodes yet...

File: _episodes/first_example_recipe.md [hunk 18]
```diff
@@ -0,0 +1,332 @@
...
+> ~~~bash
+> vim  recipe_example.yml
+> ~~~
+{: .challenge}
+
+Please note the following sections:
+  - documentation: lines 4-20
+
+    The documentation consists of the following information:
```
Comment (hb326) [inline _episodes/first_example_recipe.md:129]: This seems to be slightly doubled. A lot of information from here was already mentioned a bit earlier. We might be able to compact that a bit by, for example, only use more generic terms earlier, and not a list, or remove the detailed list from earlier and just mention the four different parts of a recipe. And then show the details when people actually look at the recipe?

File: _episodes/first_example_recipe.md [hunk 19]
```diff
@@ -0,0 +1,332 @@
...
+
+    The dataset definition consists of a list of python dictionaries with the information on the datasets.
+    - dataset name (key: dataset)
+    - project (key: project)
+    - experiment (key: exp)
+    - mip (for CMIP data, key: mip)
+    - ensemble member (key: ensemble)
+    - time range (e.g. key-value-pair: start_year: 1982, end_year: 1990)
+    - model grid (for CMIP6 data only, key: grid)
+    - alias (key: alias; use the alias for e.g. a more human readable name)
```
Comment (hb326) [inline _episodes/first_example_recipe.md:147]: Maybe rephrase to "... a more human readable name for the dataset)"

File: _episodes/first_example_recipe.md (hunk 17 above)
Comment (BenMGeo) [reply _episodes/first_example_recipe.md:120]: Something to mention at the next call @workshop.

File: _episodes/first_example_recipe.md [hunk 20]
```diff
@@ -0,0 +1,332 @@
...
+
+    The documentation consists of the following information:
+    - description: a short description of the recipe
+    - authors: a list of authors (linked to esmvaltool/config-references.yml)
+    - maintainer: a list of maintainers (linked to esmvaltool/config-references.yml)
+    - references: a list of references (linked to a bibtexfile in esmvaltool/references with the same name)
+    - projects: a list of projects (linked to esmvaltool/config-references.yml)
+
+
+  - datasets: lines 22-23
```
Comment (hb326) [inline _episodes/first_example_recipe.md:137]: Same here as above: the information here seems slightly redundant since it has been mentioned in some detail already above. So maybe shorten the description above, or remove it there...

File: _episodes/first_example_recipe.md (hunk 18 above)
Comment (BenMGeo) [reply _episodes/first_example_recipe.md:129]: Open for discussion. I think it is very important so it is worth telling twice.

File: _episodes/first_example_recipe.md (hunk 11 above)
Comment (BenMGeo) [reply _episodes/first_example_recipe.md:45]: This comes before. So do you suggest to link ep5?

**Comment by Peter9192** [comment]
Are you still working on this @BenMGeo or do you need another round of review/approval?

**Comment by BenMGeo** [comment]
> 
> 
> Are you still working on this @BenMGeo or do you need another round of review/approval?

Yes, including solutions.

**Comment by BenMGeo** [comment]
Only Links are missing, that can be linked after the merge?

**Comment by hb326** [comment]
Sorry! Wanted to review this before my vacation but did not get around to do so.

**Comment by Peter9192** [comment]
@BenMGeo shall I merge this now? Can you make a new issue about the links and possible other open ends in the conversations above?

**Comment by BenMGeo** [comment]
> 
> 
> @BenMGeo shall I merge this now? Can you make a new issue about the links and possible other open ends in the conversations above?

Thank you for checking the line length. I thought I had the editor do it automatically. Must check the settings.

We should merge it. I set up an issue for the remaining parts #64.
